firmware/Core/Src/mpi/mpi_command_handling.c

```diff
@@ -0,0 +1,118 @@
+	memset(UART1_txBuffer, 0, UART1_txBuffer_size); 	// Initialize all elements to 0
+
+	// All commands to the MPI must begin with 'TC'
+	UART1_txBuffer[0] = 0x54;	// Hex representation of 'T'
+	UART1_txBuffer[1] = 0x43;	// Hex representation of 'C'
+
+	// Add the command bytes & parameters to the Transmit Buffer
+	if (bytes_to_send_len > 0)
+	{
+		for (size_t i = 0; i < bytes_to_send_len; i++)
+		{
+			UART1_txBuffer[2 + i] = bytes_to_send[i];
+		}
+	}
+
+	// Transmit the MPI command
+	uint32_t timeout_duration_ms = 2000;	// CONFIGURATION PARAMETER //FIXME: change to 100 later
+	HAL_StatusTypeDef transmit_status = HAL_UART_Transmit(&huart1, (uint8_t *)UART1_txBuffer, UART1_txBuffer_size, timeout_duration_ms);
+
+	// Check UART transmission
+	if (transmit_status != HAL_OK)
+	{
+		return 1; // Error code: Failed UART transmission
+	}
+
+	// Receive MPI response byte by byte (This is done to account for potential errors from the mpi where it doesnt send back an expected response)
+	HAL_UART_DMAResume(&huart1);
+	HAL_StatusTypeDef receive_status = HAL_UART_Receive_DMA(&huart1, (uint8_t*) &UART_mpi_rx_last_byte, 1);
+	
+	// Check for UART reception errors
+    if (receive_status != HAL_OK) {
+        return 2; // Error code: Failed UART reception
+    }
+
+	uint32_t UART_mpi_rx_start_time_ms = HAL_GetTick();
+
+	while(1){
+		// MPI hasn't sent any data and has timed out
+		if(((HAL_GetTick() - UART_mpi_rx_start_time_ms) > timeout_duration_ms) && (UART_mpi_rx_buffer_write_idx == 0)){
+			DEBUG_uart_print_str("Break1\n");
+			break; // TODO: handle case
+		}
+		// MPI has sent some data and has timed out
+		if (((HAL_GetTick() - UART_mpi_rx_last_byte_write_time_ms) > timeout_duration_ms) && (UART_mpi_rx_buffer_write_idx > 0)) {
+			// DEBUG_uart_print_str("Size of MPI response buffer received: ");
+			// DEBUG_uart_print_int32((uint32_t)UART_mpi_rx_buffer_write_idx);
+			// DEBUG_uart_print_str("\n");
+
+			// Copy the buffer to the last received byte index
+			for (uint16_t i = 0; i < UART_mpi_rx_buffer_write_idx; i++) {
+				mpi_response[i] = UART_mpi_rx_buffer[i];
+			}
+
+			// TODO: Validate MPI response
+
+			// Clear the buffer (memset to 0, but volatile-compatible) and reset the write pointer.
+			for (uint16_t i = 0; i < UART_mpi_rx_buffer_len; i++) {
+				UART_mpi_rx_buffer[i] = 0;
+			}
+			UART_mpi_rx_buffer_write_idx = 0;
+
+			DEBUG_uart_print_str("Break2\n");
+			break; // TODO: handle case
+		}
+		// TODO: write_idx >= sizeof(rx_buffer)
+	}
+	HAL_UART_DMAStop(&huart1);
+	//HAL_StatusTypeDef receive_status = HAL_UART_Receive_DMA(&huart1, mpi_response, mpi_response_size);
+	
+	// Start the UART reception process in interrupt mode from the MPI
+    
+	// HAL_UART_Receive_DMA(&huart1, mpi_response, mpi_response_size);
+	//HAL_Delay(100);
+
+	// Check MPI response for errors: It should be an echo of the cmd sent along with an appended response code
```

Comment:
I think we may want to avoid doing this check, and instead just print out what was received regardless in the caller.

If managed correctly, returning an error here is fine, but it's imperative that that the caller still prints the received content, even in this special success-like error case.

Reply:
Was this implemented?